Segfault attempting to delete ELB

/kind bug

**What steps did you take and what happened:**

CAPA encounters a nil pointer exception while attempting to delete a cluster. Specifically, it appears to be failing while attempting to delete an ELB. Relevant logs are below. This happens on each reconciliation attempt so my pod is in crash loop backoff.

```
I0708 16:21:27.783453       1 cluster_controller.go:91] Running reconcile Cluster for "andrewwi-test2"
I0708 16:21:27.783480       1 cluster_controller.go:125] reconciling cluster object andrewwi-test2 triggers delete.
I0708 16:21:27.783505       1 actuator.go:130] [cluster-actuator] "level"=0 "msg"="Deleting cluster"  "cluster-name"="andrewwi-test2" "cluster-namespace"="c91c1f6096ba47814393c0aa8ece123e4a429117"
I0708 16:21:27.784507       1 loadbalancer.go:82] [cluster-actuator]/cluster.k8s.io/v1alpha1/c91c1f6096ba47814393c0aa8ece123e4a429117/andrewwi-test2 "level"=2 "msg"="Deleting load balancers"
I0708 16:21:27.785394       1 scope.go:217] [cluster-actuator]/cluster.k8s.io/v1alpha1/c91c1f6096ba47814393c0aa8ece123e4a429117/andrewwi-test2 "level"=1 "msg"="updating cluster status"
E0708 16:21:27.873423       1 runtime.go:69] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:76
vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
GOROOT/src/runtime/panic.go:522
GOROOT/src/runtime/panic.go:82
GOROOT/src/runtime/signal_unix.go:390
pkg/cloud/aws/services/elb/loadbalancer.go:159
pkg/cloud/aws/services/elb/loadbalancer.go:85
pkg/cloud/aws/actuators/cluster/actuator.go:146
vendor/sigs.k8s.io/cluster-api/pkg/controller/cluster/cluster_controller.go:126
vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215
vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158
vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
src/runtime/asm_amd64.s:1337
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1384008]

goroutine 152 [running]:
sigs.k8s.io/cluster-api-provider-aws/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x105
panic(0x16e7680, 0x2c77850)
        GOROOT/src/runtime/panic.go:522 +0x1b5
sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/services/elb.(*Service).getAPIServerClassicELBSpec(0xc0001bc580, 0x19252c7)
        pkg/cloud/aws/services/elb/loadbalancer.go:159 +0x3e8
sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/services/elb.(*Service).DeleteLoadbalancers(0xc0001bc580, 0xc0001bc580, 0xc000e172c0)
        pkg/cloud/aws/services/elb/loadbalancer.go:85 +0x8c
sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators/cluster.(*Actuator).Delete(0xc00007b440, 0xc0004436c0, 0x0, 0x0)
        pkg/cloud/aws/actuators/cluster/actuator.go:146 +0x35a
sigs.k8s.io/cluster-api-provider-aws/vendor/sigs.k8s.io/cluster-api/pkg/controller/cluster.(*ReconcileCluster).Reconcile(0xc0002f1aa0, 0xc0005b7a40, 0x28, 0xc00057d880, 0xe, 0x2c8c3c0, 0xc000d75ed0, 0x43da53, 0xc000d75f30)
        vendor/sigs.k8s.io/cluster-api/pkg/controller/cluster/cluster_controller.go:126 +0x70a
sigs.k8s.io/cluster-api-provider-aws/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00056d720, 0x0)
        vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215 +0x1cc
sigs.k8s.io/cluster-api-provider-aws/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1()
        vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158 +0x36
sigs.k8s.io/cluster-api-provider-aws/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0004a0860)
        vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54
sigs.k8s.io/cluster-api-provider-aws/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0004a0860, 0x3b9aca00, 0x0, 0x1, 0xc0000881e0)
        vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xf8
sigs.k8s.io/cluster-api-provider-aws/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc0004a0860, 0x3b9aca00, 0xc0000881e0)
        vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/cluster-api-provider-aws/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start
        vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:157 +0x311
```

Here is a sanitized copy of the actual cluster object I believe it is blowing up on:

```
apiVersion: cluster.k8s.io/v1alpha1
kind: Cluster
metadata:
  creationTimestamp: "2019-07-03T22:40:58Z"
  deletionGracePeriodSeconds: 0
  deletionTimestamp: "2019-07-08T16:27:10Z"
  finalizers:
  - cluster.cluster.k8s.io
  generation: 3
  name: andrewwi-test2
  namespace: c91c1f6096ba47814393c0aa8ece123e4a429117
  resourceVersion: "23357020"
  selfLink: /apis/cluster.k8s.io/v1alpha1/namespaces/c91c1f6096ba47814393c0aa8ece123e4a429117/clusters/andrewwi-test2
  uid: a064ccef-9de3-11e9-b818-0674078ea66e
spec:
  clusterNetwork:
    pods:
      cidrBlocks:
      - 192.168.0.0/16
    serviceDomain: cluster.local
    services:
      cidrBlocks:
      - 10.96.0.0/12
  providerSpec:
    value:
      caKeyPair:
        cert: NOPE
        key: NOPE
      clusterConfiguration:
        apiServer:
          extraArgs:
            enable-admission-plugins: NodeRestriction,PodSecurityPolicy
        certificatesDir: ""
        controlPlaneEndpoint: ""
        controllerManager: {}
        dns:
          type: ""
        etcd: {}
        imageRepository: ""
        kubernetesVersion: ""
        networking:
          dnsDomain: ""
          podSubnet: ""
          serviceSubnet: ""
        scheduler: {}
      etcdCAKeyPair:
        cert: NOPE
        key: NOPE
      frontProxyCAKeyPair:
        cert: NOPE
        key: NOPE
      metadata:
        creationTimestamp: null
      networkSpec:
        vpc: {}
      region: us-west-2
      saKeyPair:
        cert: NOPE
        key: NOPE
      sshKeyName: a-keypair
status:
  apiEndpoints:
  - host: ""
    port: 6443
  providerStatus:
    bastion:
      id: ""
    metadata:
      creationTimestamp: null
    network:
      apiServerElb:
        attributes: {}
```

**What did you expect to happen:**

No segfaults.

**Anything else you would like to add:**

For some further background that may help, at the time that I created the cluster the CAPA controller had incorrect credentials for my AWS account. So no resources were ever created on the Amazon side for this cluster. I later fixed the credentials and deleted the cluster. Unfortunately, I don't remember if I initiated the deletion before or after I corrected the credentials.


**Environment:**

- Cluster-api-provider-aws version: 0.3.2
- Cluster-api version: 0.1.4
- Kubernetes version: (use `kubectl version`): 1.12.5

--- a/pkg/cloud/aws/services/elb/BUILD.bazel
+++ b/pkg/cloud/aws/services/elb/BUILD.bazel
@@ -24,13 +24,20 @@ go_library(
 
 go_test(
     name = "go_default_test",
-    srcs = ["service_test.go"],
+    srcs = [
+        "loadbalancer_test.go",
+        "service_test.go",
+    ],
     embed = [":go_default_library"],
     deps = [
+        "//pkg/apis/awsprovider/v1alpha1:go_default_library",
         "//pkg/cloud/aws/actuators:go_default_library",
         "//pkg/cloud/aws/services/ec2/mock_ec2iface:go_default_library",
         "//pkg/cloud/aws/services/elb/mock_elbiface:go_default_library",
+        "//vendor/github.com/aws/aws-sdk-go/aws:go_default_library",
+        "//vendor/github.com/aws/aws-sdk-go/service/elb:go_default_library",
         "//vendor/github.com/golang/mock/gomock:go_default_library",
+        "//vendor/k8s.io/apimachinery/pkg/apis/meta/v1:go_default_library",
         "//vendor/sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1:go_default_library",
     ],
 )

--- a/pkg/cloud/aws/services/elb/loadbalancer.go
+++ b/pkg/cloud/aws/services/elb/loadbalancer.go
@@ -81,19 +81,18 @@ func (s *Service) GetAPIServerDNSName() (string, error) {
 func (s *Service) DeleteLoadbalancers() error {
 	s.scope.V(2).Info("Deleting load balancers")
 
-	// Get default api server spec
-	spec := s.getAPIServerClassicELBSpec()
+	// Get default api server name.
+	elbName := GenerateELBName(s.scope.Name(), v1alpha1.APIServerRoleTagValue)
 
-	// Describe or create.
-	apiELB, err := s.describeClassicELB(spec.Name)
-	if IsNotFound(err) {
-		return nil
-	}
-	if err != nil {
+	// Describe and delete if exists.
+	if _, err := s.describeClassicELB(elbName); err != nil {
+		if IsNotFound(err) {
+			return nil
+		}
 		return err
 	}
 
-	if err := s.deleteClassicELBAndWait(apiELB.Name); err != nil {
+	if err := s.deleteClassicELBAndWait(elbName); err != nil {
 		return err
 	}
 
@@ -137,7 +136,6 @@ func GenerateELBName(clusterName string, elbName string) string {
 }
 
 func (s *Service) getAPIServerClassicELBSpec() *v1alpha1.ClassicELB {
-
 	res := &v1alpha1.ClassicELB{
 		Name:   GenerateELBName(s.scope.Name(), v1alpha1.APIServerRoleTagValue),
 		Scheme: v1alpha1.ClassicELBSchemeInternetFacing,

--- a/pkg/cloud/aws/services/elb/loadbalancer_test.go
+++ b/pkg/cloud/aws/services/elb/loadbalancer_test.go
@@ -1,0 +1,100 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package elb
+
+import (
+	"testing"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/elb"
+	"github.com/golang/mock/gomock"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/apis/awsprovider/v1alpha1"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/actuators"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/services/ec2/mock_ec2iface"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/aws/services/elb/mock_elbiface"
+	clusterv1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+)
+
+func TestDeleteLoadBalancers(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	testCases := []struct {
+		name   string
+		expect func(m *mock_elbiface.MockELBAPIMockRecorder)
+	}{
+		{
+			name: "elb exists, it deletes it",
+			expect: func(m *mock_elbiface.MockELBAPIMockRecorder) {
+				m.DescribeLoadBalancers(gomock.Any()).Return(&elb.DescribeLoadBalancersOutput{
+					LoadBalancerDescriptions: []*elb.LoadBalancerDescription{
+						{
+							LoadBalancerName: aws.String("test-cluster-apiserver"),
+							VPCId:            aws.String("test-vpc"),
+							Scheme:           aws.String(string(v1alpha1.ClassicELBSchemeInternetFacing)),
+						},
+					},
+				}, nil)
+
+				m.DescribeLoadBalancerAttributes(gomock.Any()).Return(&elb.DescribeLoadBalancerAttributesOutput{
+					LoadBalancerAttributes: &elb.LoadBalancerAttributes{},
+				}, nil)
+
+				m.DeleteLoadBalancer(&elb.DeleteLoadBalancerInput{
+					LoadBalancerName: aws.String("test-cluster-apiserver"),
+				})
+
+				m.DescribeLoadBalancers(gomock.Any()).Return(&elb.DescribeLoadBalancersOutput{}, nil)
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			ec2Mock := mock_ec2iface.NewMockEC2API(mockCtrl)
+			elbMock := mock_elbiface.NewMockELBAPI(mockCtrl)
+
+			scope, err := actuators.NewScope(actuators.ScopeParams{
+				Cluster: &clusterv1.Cluster{
+					ObjectMeta: metav1.ObjectMeta{Name: "test-cluster"},
+				},
+				AWSClients: actuators.AWSClients{
+					EC2: ec2Mock,
+					ELB: elbMock,
+				},
+			})
+
+			scope.ClusterConfig = &v1alpha1.AWSClusterProviderSpec{
+				NetworkSpec: v1alpha1.NetworkSpec{
+					VPC: v1alpha1.VPCSpec{
+						ID: "test-vpc",
+					},
+				},
+			}
+			if err != nil {
+				t.Fatalf("Failed to create test context: %v", err)
+			}
+
+			tc.expect(elbMock.EXPECT())
+			s := NewService(scope)
+			if err := s.DeleteLoadbalancers(); err != nil {
+				t.Fatalf("got an unexpected error: %v", err)
+			}
+		})
+	}
+}

